packages/dds/cell/src/test/cell.rollback.spec.ts

```diff
@@ -0,0 +1,166 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "node:assert";
+
+import { setupRollbackTest, createAdditionalClient } from "@fluid-private/test-dds-utils";
+
+import { CellFactory } from "../cellFactory.js";
+import type { ISharedCell } from "../interfaces.js";
+
+const cellFactory = new CellFactory();
+
+describe("Cell with rollback", () => {
+	it("should emit valueChanged on set and rollback should re-emit previous value", async () => {
+		const { dds: cell, containerRuntime } = setupRollbackTest<ISharedCell>(
+			"cell-1",
+			(rt, id): ISharedCell => cellFactory.create(rt, id),
+		);
+
+		const events: (string | undefined)[] = [];
+
+		cell.on("valueChanged", (value) => events.push("valueChanged"));
+
+		cell.set(10);
+		assert.equal(cell.get(), 10);
+
+		containerRuntime.rollback?.();
+
+		assert.equal(cell.get(), undefined);
+
+		assert.deepEqual(events, ["valueChanged"]);
+	});
+
+	it("should emit delete on delete, and rollback should re-emit last valueChanged", async () => {
+		const {
+			dds: cell,
+			containerRuntimeFactory,
+			containerRuntime,
+		} = setupRollbackTest<ISharedCell>(
+			"cell-1",
+			(rt, id): ISharedCell => cellFactory.create(rt, id),
+		);
+
+		const events: (string | undefined)[] = [];
+
+		cell.on("valueChanged", (value) => events.push("valueChanged"));
+		cell.on("delete", () => events.push("delete"));
+
+		cell.set(42);
+		containerRuntime.flush();
```

Comment:
Another approach for granularity is to shift the events off as you check them
```suggestion
		cell.set(42);
		assert.equal(events.shift(), "valueChanged");
		containerRuntime.flush();
```